### PR TITLE
Cleanup semantics of Boolean python builtins

### PR DESCRIPTION
**Pull request by IgnaceBleukx**
Cleanup of Boolval-related stuff in constructors and (overloaded) operators.
- cp.any and cp.all are now guaranteed to return a CPMpy expression if any of the arguments is a CPMpy expression, even when there is a Python constant in it
- same for overloaded &, | and ^ -  now also works properly on `BoolVal`

Also added a clearer error message in the `cp.any` and `cp.all` constructors for when a user gives it a nested list.

### CODE REVIEW

File: cpmpy/expressions/core.py [hunk 1]
```diff
@@ -432,11 +432,55 @@ def __invert__(self):
...
-
+    
     def get_bounds(self):
         v = int(self.args[0])
         return (v,v)
 
+    def __and__(self, other):
+        if self.args[0]:
+            if isinstance(other, Expression): return other
+            else: return BoolVal(other) # ensure it is an Expression as result
```
Comment (ThomSerg) [inline cpmpy/expressions/core.py:443]: Will fail on integers in a boolean context:
```python
cp.BoolVal(True) & 1
cp.BoolVal(False) | 0
```
Should we add "0" and "1" to `is_true_cst`? (afterwards in the constructor for `BoolVal` it gets converted into a boolean) Or handle both in the constructor for `BoolVal`?

File: tests/test_builtins.py [hunk 2]
```diff
@@ -44,4 +45,66 @@ def test_abs(self):
...
+        x = [cp.boolvar(), False, cp.boolvar()]
+        self.assertEqual(str(cpm_all(x)), "boolval(False)")
+        x = [False, cp.BoolVal(False)]
+        self.assertEqual(str(cpm_all(x)), "boolval(False)")
+        x = [cp.BoolVal(False), False]
+        self.assertEqual(str(cpm_all(x)), "boolval(False)")
+        # only Python constants
+        x = [False, True]
+        self.assertEqual(str(cpm_all(x)), "False")
+        # only CPMPy True constant
```
Comment (ThomSerg) [inline tests/test_builtins.py:67]: "only" -> "one"
Also a test for when 2 or more CPMpy true constants? (se that we get a `Operator("and", collect)` instead of just the first element of collect

**Review by ThomSerg — COMMENTED** [review]
Just one small issue when using ints (0 or 1) in a boolean context.

File: cpmpy/expressions/core.py (hunk 1 above)
Comment (IgnaceBleukx) [reply cpmpy/expressions/core.py:443]: Aha yes idd, I'll add a type error when `other` is numerical - we do the same in the superclass

File: cpmpy/expressions/python_builtins.py [hunk 3]
```diff
@@ -32,25 +32,40 @@
...
     collect = [] # logical expressions
+    is_expr, return_false = False, False
     for elem in iterable:
-        if is_false_cst(elem):
-            return False  # no need to create constraint
+        if isinstance(elem, Expression): # probably most likely case
+            is_expr = True
+            if isinstance(elem, BoolVal):
+                if not elem.args[0]: # False constant
+                    return_false = True
```
[the remark this answers is not in the document]
Comment (IgnaceBleukx) [reply cpmpy/expressions/python_builtins.py:46]: No, it's because we don't know yet what type to return.
Either they're all Python booleans, so then we return Python bool, or we have at least one expression, and then we return `BoolVal(False)`

[the remark this answers is not in the document]
Comment (IgnaceBleukx) [reply cpmpy/expressions/python_builtins.py:46]: Fixed this and added an explicit test-case as well

**Comment by IgnaceBleukx** [comment]
Should be ready now @ThomSerg

File: cpmpy/expressions/python_builtins.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThomSerg) [reply cpmpy/expressions/python_builtins.py:46]: To return the default python `all` over only python booleans, you indeed need to check all elements to see what type. But if the array contains at least one `cp.BoolVal(False)` you could theoretically early-exit with a `cp.BoolVal(False)`, but then you don't type-check the rest of the array.

File: tests/test_builtins.py [hunk 4]
```diff
@@ -44,4 +45,83 @@ def test_abs(self):
...
+
+    def test_all(self):
+        # edge-cases
+        # CPMpy False -> should return CPMpy False
+        x = [cp.boolvar(), cp.BoolVal(False), cp.boolvar()]
+        self.assertEqual(str(cpm_all(x)), "boolval(False)")
+        # python False, but also expression! -> should return CPMpy False
+        x = [cp.boolvar(), False, cp.boolvar()]
+        self.assertEqual(str(cpm_all(x)), "boolval(False)")
+        x = [False, cp.BoolVal(False)]
```
Comment (ThomSerg) [inline tests/test_builtins.py:60]: one with `[False, BoolVal(True)]`?

**Review by ThomSerg — CHANGES_REQUESTED** [review]
Just a small copy-paste mistake in the test set

File: tests/test_builtins.py [hunk 5]
```diff
@@ -44,4 +45,83 @@ def test_abs(self):
...
+        # only Python constants
+        x = [False, True]
+        self.assertEqual(str(cpm_any(x)), "True")
+        # one CPMPy True constant
+        x = [cp.BoolVal(True)]
+        self.assertEqual(str(cpm_any(x)), "boolval(True)")
+        # one python True constant
+        x = [True]
+        self.assertEqual(str(cpm_any(x)), "True")
+        # Python and CPMpy True constant
```
[the remark this answers is not in the document]
Comment (IgnaceBleukx) [reply tests/test_builtins.py]: Resturctured the tests and added one more

**Comment by IgnaceBleukx** [comment]
After some re-trying, the tests actually worked on the Github runner... @ThomSerg fine to merge now?

**Review by ThomSerg — APPROVED** [review]
Perfect
